File: rebol/pr005.r
#! /usr/bin/env r3

REBOL [
    Title: "Problem 5"

    Description: {
         Problem 5

         30 November 2001

         2520 is the smallest number that can be divided by each of the numbers
         from 1 to 10 without any remainder.

         What is the smallest positive number that is evenly divisible by all of
         the numbers from 1 to 20?

    }

    Solution: 232792560
]

gcd: func [a [integer!] b [integer!] /local tmp] [
    while [b <> 0] [
	tmp: b
	b: a // b
	a: tmp
    ]
    a
]

lcm: func [a [integer!] b [integer!]] [
    a / (gcd a b) * b
]

solve: funct [] [
    total: 1
    for i 2 20 1 [
	total: lcm total i
    ]
    total
]

print solve

File: zig/pr005.zig
// Problem 5
//
// 30 November 2001
//
// 2520 is the smallest number that can be divided by each of the numbers
// from 1 to 10 without any remainder.
//
// What is the smallest positive number that is evenly divisible by all of
// the numbers from 1 to 20?

// define_problem!(pr005, 5, 232792560);

const std = @import("std");
const misc = @import("misc.zig");
const assert = std.debug.assert;
const print = std.debug.print;

pub fn main() !void {
    var accum: u64 = 1;

    var i: u64 = 2;
    while (i <= 20) : (i += 1) {
        accum = misc.lcm(accum, i);
    }

    assert(accum == 232792560);
    print("5 {d}\n", .{accum});
}
